sweepbatcher/sweep_batch.go

```diff
@@ -0,0 +1,1390 @@
+	return &batch{
+		id:               bk.id,
+		state:            bk.state,
+		primarySweepID:   bk.primaryID,
+		sweeps:           bk.sweeps,
+		requestChan:      make(chan *sweep),
+		returnChan:       bk.returnChan,
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		confChan:         make(chan *chainntnfs.TxConfirmation),
+		errChan:          make(chan error, 1),
+		callEnter:        make(chan struct{}),
+		callLeave:        make(chan struct{}),
+		quit:             make(chan struct{}),
+		completeChan:     bk.completeChan,
+		batchTxid:        bk.batchTxid,
+		batchPkScript:    bk.batchPkScript,
+		rbfCache:         bk.rbfCache,
+		wallet:           bk.wallet,
+		chainNotifier:    bk.chainNotifier,
+		signerClient:     bk.signerClient,
+		muSig2SignSweep:  bk.musig2SignSweep,
+		verifySchnorrSig: bk.verifySchnorrSig,
+		store:            bk.store,
+		log:              bk.log,
+		Cfg:              &cfg,
+	}
+}
+
+// handleSweep adds a sweep to the batch. If this is the first sweep being added
+// to the batch then it also sets the primary sweep ID.
+func (b *batch) handleSweep(ctx context.Context, sw sweep) error {
+	if b.primarySweepID == lntypes.ZeroHash {
+		b.primarySweepID = sw.swapHash
+		b.Cfg.BatchConfTarget = sw.confTarget
+
+		// If this is the first sweep being added to the batch, we also
+		// need to start the spend monitor for this new primary sweep.
+		b.monitorSpend(ctx, sw)
+	}
+
+	// Since all the actions of the batch happen sequentially, we could
+	// arrive here after the batch got closed because of a spend. In this
+	// case we must return the sweep request back to the batcher and return
+	// from this function. This is an expected behavior and should not
+	// return any error.
+	if b.state != Open {
+		sweepReq := SweepRequest{
+			SwapHash: sw.swapHash,
+			Outpoint: sw.outpoint,
+			Value:    sw.value,
+			Notifier: sw.notifier,
+		}
+		go purgeSweep(ctx, sweepReq, b.returnChan)
+		return nil
+	}
+
+	if b.primarySweepID == sw.swapHash {
+		b.Cfg.BatchConfTarget = sw.confTarget
+	}
+
+	_, ok := b.sweeps[sw.swapHash]
+	if !ok {
+		b.log.Infof("adding sweep %x", sw.swapHash[:6])
+	}
+
+	b.sweeps[sw.swapHash] = sw
+
+	return b.persistSweep(ctx, sw, false)
+}
+
+// sweepExists returns true if the batch contains the sweep with the given hash.
+func (b *batch) sweepExists(hash lntypes.Hash) bool {
+	done, err := b.scheduleNextCall()
+	defer done()
+	if err != nil {
+		return false
+	}
+
+	_, ok := b.sweeps[hash]
+
+	return ok
+}
+
+// acceptsSweep returns true if the batch is able to accept the given sweep. To
+// accept a sweep a batch needs to be in an Open state, and the incoming sweep's
+// timeout must not be too far away from the other timeouts of existing sweeps.
+// If this batch contains a single sweep that spends to a non-wallet address we
+// are not accepting other sweeps as batching is pointless in this case.
+func (b *batch) acceptsSweep(sweep *sweep) bool {
```

Comment:
> It'd be slightly better if from the Batcher's point of view, we could only call addSweep with a new or existing sweep and addSweep would hook into the batch's event loop with scheduleNextCall and then:

The `sweepExists` and `addSweep` need to be separate, as you first need to scan all the batches to see if a sweep is already part of a batch before attempting to add it to another one. This way you avoid situations where the sweep is in 2 different batches, which would lead to tx conflict.

> The reason to do so is otherwise by the time the Batcher gets a positive answer from acceptSweep the batch might already be done. It's a bit nicer imo to handle the whole thing in one atomic step.

Yes, by the time we attempt to add the sweep the batch could have been confirmed and even have signaled its exit. But keep in mind that the batcher is responsible for the actual deletion of the batch, and the above procedure of handling the sweep request is a task of the **batcher**. This means that between every `sweepExists` / `acceptsSweep` / `addSweep` no batch could have been deleted, as the batcher would have to return to its main run loop to handle a batch deletion.

Reply:
> > It'd be slightly better if from the Batcher's point of view, we could only call addSweep with a new or existing sweep and addSweep would hook into the batch's event loop with scheduleNextCall and then:
> 
> The `sweepExists` and `addSweep` need to be separate, as you first need to scan all the batches to see if a sweep is already part of a batch before attempting to add it to another one. This way you avoid situations where the sweep is in 2 different batches, which would lead to tx conflict.
> 
> > The reason to do so is otherwise by the time the Batcher gets a positive answer from acceptSweep the batch might already be done. It's a bit nicer imo to handle the whole thing in one atomic step.
> 
> Yes, by the time we attempt to add the sweep the batch could have been confirmed and even have signaled its exit. But keep in mind that the batcher is responsible for the actual deletion of the batch, and the above procedure of handling the sweep request is a task of the **batcher**. This means that between every `sweepExists` / `acceptsSweep` / `addSweep` no batch could have been deleted, as the batcher would have to return to its main run loop to handle a batch deletion.

I think you can just scan and try to add, it'd be atomic that way.

Reply:
So something like this:

```
func (b *batch) addSweep(sweep *sweep) (bool, error) {
	done, err := b.scheduleNextCall()
	defer done()
	if err != nil {
		return false, err
	}

	// Check if the batch already has this sweep.
	_, ok := b.sweeps[sweep.swapHash]
	if ok {
		return true, nil
	}

	// Now check if we can accept it.
	if b.state != Open {
		return false, nil
	}

	for _, s := range b.sweeps {
		if s.isExternalAddr || sweep.isExternalAddr {
			return false, nil
		}
	}

	for _, s := range b.sweeps {
		timeoutDistance :=
			int32(math.Abs(float64(sweep.timeout - s.timeout)))

		if timeoutDistance > b.Cfg.MaxTimeoutDistance {
			return false, nil
		}
	}

	// Finally add the sweep.
	if err := b.handleSweep(b.runCtx, *sweep); err != nil {
		return false, err

	}

	return true, nil
}
```

Reply:
> // Check if the batch already has this sweep.
	_, ok := b.sweeps[sweep.swapHash]
	if ok {
		return true, nil
	}

What if this whole func is running on batch1, but batch2 has the sweep? You have to first ask all the batches whether they have it, and then start attempting to add

Reply:
Yeah you're right, merging just accept and add will fix this and the addition will then be atomic.